### PR TITLE
✨Experiment✨ Extend custom DNS with AD & Tracking Protection

### PR DESCRIPTION
**Pull request by strseb**
This is based on #902 

With our Code for custom dns in place, we can extend the feature to offer ad-blocking and/or tracking blocking dns servers. 
Which i think would be nice :) 

Unbenannt

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/VPN-796)

### CODE REVIEW

File: src/ui/settings/ViewProtections.qml [hunk 1]
```diff
@@ -0,0 +1,192 @@
+
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (bakulf) [inline src/ui/settings/ViewProtections.qml:3]: Remove these 2 empty lines

File: src/ui/settings/ViewSettingsMenu.qml [hunk 2]
```diff
@@ -132,6 +132,15 @@ VPNFlickable {
             onClicked: settingsStackView.push("../settings/ViewNotifications.qml")
             visible: VPNFeatureList.captivePortalNotificationSupported || VPNFeatureList.unsecuredNetworkNotificationSupported || VPNFeatureList.notificationControlSupported
         }
+        VPNSettingsItem {
```
Comment (bakulf) [inline src/ui/settings/ViewSettingsMenu.qml:135]: I don't review the UI because we need to apply the Peter's UX when ready.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/platforms/windows/windowsnetworkwatcher.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@ void WindowsNetworkWatcher::initialize() {
 
   DWORD prefNotifSource;
   if (WlanRegisterNotification(m_wlanHandle, WLAN_NOTIFICATION_SOURCE_MSM,
-                               true /* ignore duplicates */, wlanCallback, this,
+                               true /* ignore duplicates */, (WLAN_NOTIFICATION_CALLBACK) wlanCallback, this,
```
Comment (strseb) [inline src/platforms/windows/windowsnetworkwatcher.cpp:43]: This compile error popped out of nowhere. 
msvc was not able to cast wlanCallback automatically?

**Comment by strseb** [comment]
@bakulf , @lesleyjanenorton  - I've updated the pr, and pulled new UI from #1530 into it. Only thing left are the strings.

File: src/settingsholder.h [hunk 4]
```diff
@@ -47,13 +42,26 @@ class SettingsHolder final : public QObject {
...
   ~SettingsHolder();
 
   static SettingsHolder* instance();
 
+  enum DnsProvider {
+    Custom = 4,
+    BlockTracking = 3,
+    BlockAds = 2,
+    BlockAll = 1,
+    Gateway = 0,
```
Comment (bakulf) [inline src/settingsholder.h:61]: Can you sort by values? Gateway as the first one and so on?

File: src/ui/settings/ViewAdvancedDNSSettings.qml [hunk 5]
```diff
@@ -16,229 +16,100 @@ import telemetry 0.15
 
 
 Item {
+    property string settingId_default: "default"
```
Comment (bakulf) [inline src/ui/settings/ViewAdvancedDNSSettings.qml:19]: Why do we need this? Can we simply use the Q_ENUM entries?

**Review by bakulf — CHANGES_REQUESTED** [review]
Fix the strings. Now the string PR is merged.

File: src/ui/settings/ViewAdvancedDNSSettings.qml (hunk 5 above)
Comment (strseb) [reply src/ui/settings/ViewAdvancedDNSSettings.qml:19]: We can't do this, at least if i tried qml complained that we cant use script values as properties of a listmodel.

Comment (strseb) [reply src/ui/settings/ViewAdvancedDNSSettings.qml:19]: Update: we __could__ if we are okay with creating the list model dynamicly

Comment (lesleyjanenorton) [reply src/ui/settings/ViewAdvancedDNSSettings.qml:19]: We can get around this by sending only the string IDs in the ListModel and then `text: VPNl18n.tr(VPNl18n[tabLabelStringId])` to the button delegate in `VPNTabNavigation { }`. But this is working too!

File: src/ui/settings/ViewAdvancedDNSSettings.qml [hunk 6]
```diff
@@ -16,229 +16,97 @@ import telemetry 0.15
...
+
+    }
+
     VPNMenu {
         id: menu
         objectName: "settingsAdvancedDNSSettingsBackButton"
 
-        //% "Advanced DNS settings"
-        title: qsTrId("vpn.settings.advancedDNSSettings.title2")
+        title: "DNS Settings"
```
Comment (lesleyjanenorton) [inline src/ui/settings/ViewAdvancedDNSSettings.qml:47]: Needs a localized string.

**Review by lesleyjanenorton — APPROVED** [review]
Looks great. Left a comment with a workaround for the strings but not sure it's worth the refactor since this is working.

File: src/controller.cpp [hunk 7]
```diff
@@ -685,11 +670,10 @@ QList<IPAddressRange> Controller::getAllowedIPAddressRanges(
...
   }
   if (shouldExcludeDns()) {
+    auto dns = SettingsHolder::instance()->getDNS(server.ipv4Gateway());
     // Filter out the Custom DNS Server, if the User has one.
-    logger.debug() << "Filtering out the DNS address"
-                   << SettingsHolder::instance()->userDNS();
-    excludeIPv4s.append(
-        IPAddress::create(SettingsHolder::instance()->userDNS()));
+    logger.debug() << "Filtering out the DNS address" << dns;
+    excludeIPv4s.append(IPAddress::create(dns));
```
Comment (bakulf) [inline src/controller.cpp:676]: Are we actually sure we want to do this? Can we do the DNS lookup via VPN?

File: src/controller.cpp [hunk 8]
```diff
@@ -737,7 +721,8 @@ bool Controller::shouldExcludeDns() {
   if (!FeatureList::instance()->userDNSSupported()) {
     return false;
   }
-  if (settings->useGatewayDNS()) {
+  // Only a Custom DNS might require to be routed outside of the VPN-Tunnel
+  if (settings->dnsProvider() != SettingsHolder::DnsProvider::Custom) {
```
Comment (bakulf) [inline src/controller.cpp:725]: Is it really needed? Can we do this only for LAN servers?

File: src/ui/components/VPNViewDNSSettings.qml [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+          case settingId_adblock.toString():
+            return VPNSettings.dnsProvider === VPNSettings.BlockAds;  
+          case settingId_antiTracking.toString():
+            return VPNSettings.dnsProvider === VPNSettings.BlockTracking;  
+          case settingId_adblockAndAntiTracking.toString():
+            return VPNSettings.dnsProvider === VPNSettings.BlockAll;  
+          case settingId_local.toString():
+            return VPNSettings.dnsProvider === VPNSettings.Custom;  
+          default:
+            console.log("TODO")
```
Comment (bakulf) [inline src/ui/components/VPNViewDNSSettings.qml:38]: TODO what? :) We should log this and return ... false?

File: src/ui/components/VPNViewDNSSettings.qml [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+          case settingId_adblock.toString():
+            return VPNSettings.dnsProvider = VPNSettings.BlockAds;  
+          case settingId_antiTracking.toString():
+            return VPNSettings.dnsProvider = VPNSettings.BlockTracking;  
+          case settingId_adblockAndAntiTracking.toString():
+            return VPNSettings.dnsProvider = VPNSettings.BlockAll;  
+          case settingId_local.toString():
+            return VPNSettings.dnsProvider = VPNSettings.Custom;  
+          default:
+            console.log("TODO")
```
Comment (bakulf) [inline src/ui/components/VPNViewDNSSettings.qml:59]: No TODO, log it and set it to Gateway. What do you think?

File: src/ui/settings/ViewAdvancedDNSSettings.qml [hunk 11]
```diff
@@ -16,229 +16,97 @@ import telemetry 0.15
 
 
 Item {
+    property string settingId_default: "default"
+    property string settingId_local: "local"
+    property string settingId_adblock: "adblock"
+    property string settingId_antiTracking: "antiTracking"
+    property string settingId_adblockAndAntiTracking: "adblockAndAntiTracking"
```
Comment (bakulf) [inline src/ui/settings/ViewAdvancedDNSSettings.qml:23]: So we keep them. Ok!

File: src/ui/settings/ViewAdvancedDNSSettings.qml [hunk 12]
```diff
@@ -16,229 +16,97 @@ import telemetry 0.15
...
-
-                        VPNTextBlock {
-                           //% "Resolve website domain names using a DNS in your local network"
-                           text: qsTrId("vpn.advancedDNSSettings.localDNS.resolveWebsiteDomainNames")
-                           width: parent.width
-                        }
-                    }
+                Component.onCompleted: {
+                    defaultTabListModel.append({
+                                                "settingId": "default",
```
Comment (bakulf) [inline src/ui/settings/ViewAdvancedDNSSettings.qml:77]: Can this be set to VPNSettings.Gateway ?

**Review by bakulf — APPROVED** [review]

**Review by bakulf — APPROVED** [review]

**Pull request merged**